src/ert/config/observations.py

```diff
@@ -1,471 +1,241 @@
+
+    def __len__(self) -> int:
+        return len(self.summary_keys)
+
+    def to_xarray(self) -> xr.Dataset:
+        return (
+            pd.DataFrame(
+                data={
+                    "name": self.summary_keys,
+                    "obs_name": self.obs_names,
+                    "time": self.times,
+                    "observations": self.observations,
+                    "std": self.stds,
+                },
+            )
+            .set_index(["name", "obs_name", "time"])
+            .to_xarray()
+        )
+
+
+@dataclass
+class _GenObsDataset:
+    gen_data_keys: List[str] = field(default_factory=list)
+    observations: List[float] = field(default_factory=list)
+    stds: List[float] = field(default_factory=list)
+    indexes: List[int] = field(default_factory=list)
+    report_steps: List[int] = field(default_factory=list)
+    obs_names: List[str] = field(default_factory=list)
+
+    def __len__(self) -> int:
+        return len(self.gen_data_keys)
+
+    def to_xarray(self) -> xr.Dataset:
+        return (
+            pd.DataFrame(
+                data={
+                    "name": self.gen_data_keys,
+                    "obs_name": self.obs_names,
+                    "report_step": self.report_steps,
+                    "index": self.indexes,
+                    "observations": self.observations,
+                    "std": self.stds,
+                }
+            )
+            .set_index(["name", "obs_name", "report_step", "index"])
+            .to_xarray()
+        )
+
+
+class _GenObsAccumulator:
+    def __init__(self) -> None:
+        self.ds: _GenObsDataset = _GenObsDataset()
+
+    def write(
+        self,
+        gen_data_key: str,
+        obs_name: str,
+        report_step: int,
+        observations: List[float],
+        stds: List[float],
+        indexes: List[int],
+    ) -> None:
+        self.ds.gen_data_keys.extend([str(gen_data_key)] * len(observations))
+        self.ds.obs_names.extend([str(obs_name)] * len(observations))
+        self.ds.report_steps.extend([report_step] * len(observations))
+
+        self.ds.observations.extend(observations)
+        self.ds.stds.extend(stds)
+        self.ds.indexes.extend(indexes)
+
+
+class _SummaryObsAccumulator:
+    def __init__(self) -> None:
+        self.ds: _SummaryObsDataset = _SummaryObsDataset()
+
+    def write(
+        self,
+        summary_key: str,
+        obs_names: List[str],
+        observations: List[float],
+        stds: List[float],
+        times: List[Union[int, datetime]],
+    ) -> None:
+        self.ds.summary_keys.extend([str(summary_key)] * len(obs_names))
+        self.ds.obs_names.extend(map(str, obs_names))
+        self.ds.observations.extend(observations)
+        self.ds.stds.extend(stds)
+        self.ds.times.extend(times)
+
+
+# Columns used to form a key for observations of the response_type
+ObservationsIndices = {"summary": ["time"], "gen_data": ["index", "report_step"]}
```

Comment:
Maybe a property of `GenObservation` and `SumamryObservation`?

Reply:
Tried to address these comments by making it more generic, taking in only a primary key as an argument:
https://github.com/equinor/ert/pull/7563/commits/790621286bd2d6c78e2f56624cb4916945e24412
(will fix type issues)